File: C++/Data Structures/Fenwick Tree.cpp
class Fenwick {
public:
	int n;  vector<ll> t; int LOGN;
	Fenwick_Tree(int sz) {
		this -> n = sz;
		this -> LOGN = 31 - __builtin_clz(sz);
		t.assign(n + 5, 0);
	};
	void build(vector<ll> &a) {
	    for (int i = 1; i <= n; i++) {
			t[i] += a[i-1];
	        int par = i + (i & -i);
	        if (par <= n) t[par] += t[i];
	    }
	}
	int kth_element(int k) {
	    int idx = 0;
	    for (int i = 1 << LOGN; i > 0; i >>= 1) {
	        if (idx + i <= n && tree[idx + i] < k) {
	            idx += i;
	            k -= tree[idx];
	        }
	    }
    	return idx + 1;
	}
	ll make_query(ll idx) {
		ll res = 0;
		for (; idx > 0; idx -= (idx & (-idx)))  res += t[idx];
		return res;
	}
	ll query(ll l, ll r) {
		return make_query(r) - make_query(l - 1);
	}
	void add(ll idx, ll x) {
		for (; idx <= n; idx += (idx & (-idx)))  t[idx] += x;
	}
};

File: Python/Data Structures/Fenwick Tree.py
def query(i):
  res = 0
  while i>0:
    res += t[i]
    i -= (i&-i)
  return res

def update(i,x):
  while i<len(t):
    t[i] += x
    i += (i&-i)

def lower_bound(w):
  x = 0
  n = len(t) -1
  k = 1 << (n.bit_length() -1)
  while k:
    if x+k <= n and t[x+k] < w:
      w -= t[x+k]
      x += k
    k >>=1
  return x+1

t = [0]*(n+1)
for i in range(n): t[i+1] = a[i]
for i in range(1, n):
  p = i + (i&-i)
  if p<=n:
    t[p] = t[p] + t[i]
